src/enclave/Enclave/Aggregate.cpp

```diff
@@ -5,116 +5,38 @@
-  uint8_t **output_rows, size_t *output_rows_length) {
-
-  FlatbuffersAggOpEvaluator agg_op_eval(agg_op, agg_op_length);
-  RowReader r(BufferRefView<tuix::EncryptedBlocks>(input_rows, input_rows_length));
-  RowReader next_partition_first_row_reader(
-    BufferRefView<tuix::EncryptedBlocks>(
-      next_partition_first_row, next_partition_first_row_length));
-  RowReader prev_partition_last_group_reader(
-    BufferRefView<tuix::EncryptedBlocks>(
-      prev_partition_last_group, prev_partition_last_group_length));
-  RowReader prev_partition_last_row_reader(
-    BufferRefView<tuix::EncryptedBlocks>(
-      prev_partition_last_row, prev_partition_last_row_length));
-  RowWriter w;
-
-  if (next_partition_first_row_reader.num_rows() > 1) {
-      throw std::runtime_error(
-          std::string("Incorrect number of starting rows from next partition passed: expected 0 or 1, got ")
-          + std::to_string(next_partition_first_row_reader.num_rows()));
-  }
-  if (prev_partition_last_group_reader.num_rows() > 1) {
-      throw std::runtime_error(
-          std::string("Incorrect number of ending groups from prev partition passed: expected 0 or 1, got ")
-          + std::to_string(prev_partition_last_group_reader.num_rows()));
-  }
-  if (prev_partition_last_row_reader.num_rows() > 1) {
-      throw std::runtime_error(
-          std::string("Incorrect number of ending rows from prev partition passed: expected 0 or 1, got ")
-          + std::to_string(prev_partition_last_row_reader.num_rows()));
-  }
-
-  const tuix::Row *next_partition_first_row_ptr =
-    next_partition_first_row_reader.has_next() ? next_partition_first_row_reader.next() : nullptr;
-  agg_op_eval.set(prev_partition_last_group_reader.has_next() ?
-                  prev_partition_last_group_reader.next() : nullptr);
-  const tuix::Row *prev_partition_last_row_ptr =
-    prev_partition_last_row_reader.has_next() ? prev_partition_last_row_reader.next() : nullptr;
 
-  FlatbuffersTemporaryRow prev, cur(prev_partition_last_row_ptr), next;
-  bool stop = false;
-  if (r.has_next()) {
-    next.set(r.next());
-  } else {
-    stop = true;
-  }
-  while (!stop) {
-    // Populate prev, cur, next to enable lookbehind and lookahead
-    prev.set(cur.get());
-    cur.set(next.get());
-    if (r.has_next()) {
-      next.set(r.next());
-    } else {
-      next.set(next_partition_first_row_ptr);
-      stop = true;
-    }
-
-    if (prev.get() != nullptr && !agg_op_eval.is_same_group(prev.get(), cur.get())) {
-      agg_op_eval.reset_group();
-    }
-    agg_op_eval.aggregate(cur.get());
-
-    // Output the current aggregate if it is the last aggregate for its run
-    if (next.get() == nullptr || !agg_op_eval.is_same_group(cur.get(), next.get())) {
-      w.append(agg_op_eval.evaluate());
-    }
+  // Skip outputting the final row if the number of input rows is 0 AND
+  // 1. It's a grouping aggregation, OR
+  // 2. It's a global aggregation, the mode is final
```

Comment:
Do we need to skip outputting the final row in case 2? If I understand correctly, that can only occur if there are zero partitions in the partial stage. Not sure if that's possible, but if it is, global aggregation should still return a row.

If we don't need to skip in case 2, it would greatly simplify the condition and comment.

Reply:
If the second statement isn't added, then the global aggregation test fails when there are multiple partitions, since it will output a row for every empty partition as well:

```[info] - global aggregate - encrypted *** FAILED *** (433 milliseconds)
[info]   Array([0], [256], [0]) did not equal Array([256]) (OpaqueOperatorTests.scala:90)```